### PR TITLE
Add more tests to ASSERT_STATUS_CHECKED (4)

### PR DESCRIPTION
**Pull request by adamretter**
Fourth batch of adding more tests to ASSERT_STATUS_CHECKED.

* db_range_del_test
* db_write_test
* random_access_file_reader_test
* merge_test
* external_sst_file_test
* write_buffer_manager_test
* stringappend_test
* deletefile_test

### CODE REVIEW

File: db/external_sst_file_test.cc [hunk 1]
```diff
@@ -47,8 +47,8 @@ class ExternSSTFileLinkFailFallbackTest
       : DBTestBase("/external_sst_file_test", /*env_do_fsync=*/true),
         test_env_(new ExternalSSTTestEnv(env_, true)) {
     sst_files_dir_ = dbname_ + "/sst_files/";
-    DestroyDir(env_, sst_files_dir_);
-    env_->CreateDir(sst_files_dir_);
+    assert(DestroyDir(env_, sst_files_dir_).ok());
+    assert(env_->CreateDir(sst_files_dir_).ok());
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/external_sst_file_test.cc:51]: I will switch to `EXPECT_EQ`, but we can't use `ASSERT_`* here as those macros return values.

File: db/deletefile_test.cc [hunk 2]
```diff
@@ -210,7 +211,8 @@ TEST_F(DeleteFileTest, PurgeObsoleteFilesTest) {
   Iterator *itr = nullptr;
   CreateTwoLevels();
   itr = db_->NewIterator(ReadOptions());
-  db_->CompactRange(compact_options, &first_slice, &last_slice);
+  ASSERT_OK(itr->status());
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/deletefile_test.cc:214]: Will do.

File: db/db_impl/db_impl.cc [hunk 3]
```diff
@@ -4403,23 +4403,17 @@ Status DBImpl::IngestExternalFiles(
...
         args[0].files_checksum_func_names, next_file_number, super_version);
-    exec_results[0].first = true;
-    CleanupSuperVersion(super_version);
-  }
-  for (const auto& exec_result : exec_results) {
-    if (!exec_result.second.ok()) {
-      status = exec_result.second;
-      break;
+    if (!es.ok()) {
+      status = es;
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_impl/db_impl.cc:4428]: The part directly above is for the first status. Note the `[0]` bit... does that make sense?

Also note that from `[1]` to `[1+n]` in the loop further above, only the first non-OK status is captured.

File: db/forward_iterator.cc [hunk 4]
```diff
@@ -242,6 +242,12 @@ ForwardIterator::ForwardIterator(DBImpl* db, const ReadOptions& read_options,
 
 ForwardIterator::~ForwardIterator() {
   Cleanup(true);
+
+  // this is a local aggregation of the status of the
+  // immutable Iterators. We have to PermitUncheckedError
+  // before it is free'd otherwise ASSERT_STATUS_CHECKED
+  // will fail.
+  immutable_status_.PermitUncheckedError();
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/forward_iterator.cc:250]: Yes, thanks... I have now moved it

File: db/db_impl/db_impl_write.cc [hunk 5]
```diff
@@ -566,14 +566,20 @@ Status DBImpl::PipelinedWriteImpl(const WriteOptions& write_options,
...
     write_thread_.EnterAsMemTableWriter(&w, &memtable_write_group);
     if (memtable_write_group.size > 1 &&
         immutable_db_options_.allow_concurrent_memtable_write) {
       write_thread_.LaunchParallelMemTableWriters(&memtable_write_group);
+
+      // The status will never be written or read and this
+      // is needed to avoid ASSERT_STATUS_CHECKED failure
+      // when the memtable_write_group is free'd
+      memtable_write_group.status.PermitUncheckedError();
+
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_impl/db_impl_write.cc]: I felt that this was likely specific to just how it is used here... What do you think?

File: db/merge_test.cc [hunk 6]
```diff
@@ -81,7 +81,7 @@ std::shared_ptr<DB> OpenDb(const std::string& dbname, const bool ttl = false,
   options.merge_operator = std::make_shared<CountMergeOperator>();
   options.max_successive_merges = max_successive_merges;
   Status s;
-  DestroyDB(dbname, Options());
+  DestroyDB(dbname, Options()).PermitUncheckedError();
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/merge_test.cc]: It always returns OK. I have fixed it to use EXPECT_OK as ASSERT_OK can't be used here.

File: file/random_access_file_reader.cc [hunk 7]
```diff
@@ -221,11 +221,19 @@ Status RandomAccessFileReader::MultiRead(const IOOptions& opts,
...
       // Align and merge the read requests.
       size_t alignment = file_->GetRequiredBufferAlignment();
-      aligned_reqs.push_back(Align(read_reqs[0], alignment));
-      for (size_t i = 1; i < num_reqs; i++) {
+      for (size_t i = 0; i < num_reqs; i++) {
         const auto& r = Align(read_reqs[i], alignment);
-        if (!TryMerge(&aligned_reqs.back(), r)) {
+        if (i == 0) {
+          // head
+          aligned_reqs.push_back(r);
```
Comment (siying) [inline file/random_access_file_reader.cc:228]: Moving this into the loop doesn't seem to be relevant to this PR. Whether it makes it easier to read is arguable and the performance won't change much but doesn't seem to improve much either. So I would suggest we separate this change.

**Review by siying — APPROVED** [review]

File: file/random_access_file_reader.cc (hunk 7 above)
Comment (pdillinger) [reply file/random_access_file_reader.cc:228]: Seems to be for narrowing down the case requiring `r.status.PermitUncheckedError();`
